pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,951 @@
+		c.endpointSliceListerSynced,
+	}
+	if c.egressEnabled {
+		cacheSyncs = append(cacheSyncs, c.egressListerSynced)
+	}
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	go wait.NonSlidingUntil(func() {
+		if err := c.watchSecretChanges(stopCh); err != nil {
+			klog.ErrorS(err, "Watch Secret error", "secret", types.BGPPolicySecretName)
+		}
+	}, time.Second*10, stopCh)
+	go wait.Until(c.worker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	_, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(dummyKey)
+
+	if err := c.syncBGPPolicy(); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until another change happens.
+		c.queue.Forget(dummyKey)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		c.queue.AddRateLimited(dummyKey)
+		klog.ErrorS(err, "Syncing BGPPolicy failed, requeue")
+	}
+	return true
+}
+
+func (c *Controller) getBGPPolicy() *v1alpha1.BGPPolicy {
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	var oldestBP *v1alpha1.BGPPolicy
+	for _, bp := range allBPs {
+		if c.matchedCurrentNode(bp) {
+			if oldestBP == nil || bp.CreationTimestamp.Before(&oldestBP.CreationTimestamp) {
+				oldestBP = bp
+			}
+		}
+	}
+	return oldestBP
+}
+
+func (c *Controller) syncBGPPolicy() error {
+	startTime := time.Now()
+	defer func() {
+		klog.InfoS("Finished syncing BGPPolicy", "durationTime", time.Since(startTime))
+	}()
+
+	// Get the oldest BGPPolicy applied to the current Node as the effective BGPPolicy.
+	effectiveBP := c.getBGPPolicy()
+
+	// When the effective BGPPolicy is nil, it means that there is no available BGPPolicy.
+	if effectiveBP == nil {
+		// If the BGPPolicy state is nil, just return.
+		if c.bgpPolicyState == nil {
+			return nil
+		}
+
+		// If the BGPPolicy state is not nil, stop the BGP server and reset the state to nil, then return.
+		if err := c.bgpPolicyState.bgpServer.Stop(c.ctx); err != nil {
+			return err
+		}
+		c.bgpPolicyState = nil
+		return nil
+	}
+
+	klog.V(2).InfoS("Syncing BGPPolicy", "BGPPolicy", effectiveBP)
+	// Retrieve the listen port, local AS number and router ID from the effective BGPPolicy, and update them to the
+	// current state.
+	routerID, err := c.getRouterID()
+	if err != nil {
+		return err
+	}
+	var listenPort int32
+	if effectiveBP.Spec.ListenPort != nil {
+		listenPort = *effectiveBP.Spec.ListenPort
+	} else {
+		listenPort = defaultBGPListenPort
+	}
+	localASN := effectiveBP.Spec.LocalASN
+
+	// If the BGPPolicy state is nil, a new BGP server should be started, initialize the BGPPolicy state to store the
+	// latest BGP server instance, the listen port, the local ASN, and the router ID.
+	// If the BGPPolicy is not nil, any of the listen port, local AS number, or router ID have changed, stop the stale
```

Comment:
Nit: the stale BGP server -> the BGP server

Reply:
Done